### ISSUE TITLE
aws_eks_node_group fail when launch template is updated

### ISSUE THREAD

**Issue by cohandv** [terraform-providers/terraform-provider-aws#15007]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

`Terraform v0.12.28
+ provider.aws v3.3.0
+ provider.external v1.2.0

Your version of Terraform is out of date! The latest version
is 0.13.1. You can update by downloading from https://www.terraform.io/downloads.html`

### Affected Resource(s)

* aws_eks_node_group

### Terraform Configuration Files

```hcl
resource "aws_launch_template" "main" {
  dynamic "block_device_mappings" {
    content {
      device_name = "/dev/xvda"
      ebs {
        volume_size = var.root_volume_size
      }
    }
    for_each = var.root_volume_size != null ? [1] : []
  }

  dynamic "block_device_mappings" {
    content {
      device_name = volume.value["device"]
      ebs {
        delete_on_termination = lookup(volume.value, "delete_on_termination", true)
        iops                  = lookup(volume.value, "iops", null)
        snapshot_id           = lookup(volume.value, "snapshot_id", null)
        volume_size           = lookup(volume.value, "size", 1)
        volume_type           = lookup(volume.value, "type", "gp2")
      }
    }
    for_each = values(data.external.volume)[*].result
    iterator = volume
  }

  image_id      = module.ami.id
  instance_type = var.instance_type

  lifecycle {
    create_before_destroy = true
  }

  name = local.node_group_name
  tags = local.tags

  tag_specifications {
    resource_type = "instance"
    tags          = local.tags
  }

  tag_specifications {
    resource_type = "volume"
    tags          = local.tags
  }

  user_data              = local.user_data
  vpc_security_group_ids = local.security_group_ids
}


resource "aws_eks_node_group" "worker" {
  cluster_name = var.cluster_name

  labels = var.labels
  launch_template {
    id      = aws_launch_template.main.id
    version = aws_launch_template.main.latest_version
  }

  node_group_name = local.node_group_name
  node_role_arn   = module.role.arn

  scaling_config {
    desired_size = var.desired_size
    max_size     = var.max_size
    min_size     = var.min_size
  }

  subnet_ids = module.private_subnets.ids
  tags       = local.tags
}
```

### Debug Output

https://gist.github.com/cohandv/af8fd38d5b4422ec2f1ff354fae8cf85

### Expected Behavior

It should have created a new version of the launch template, and updated the node group to use latest version

### Actual Behavior

it fails saying that user data changed (it actually did not change that)

### Steps to Reproduce

1. Create a basic LT, eks node group
2. `terraform apply`
3. Change the list of security groups on the LT
4. `terraform apply`

**Comment by gdavison** [comment]
Hi @cohandv. The error message that I'm seeing says that `User data was not in the MIME multipart format`. Would you be able to include the user data in this issue, please? If it contains sensitive data, you can mask that.

**Comment by DrFaust92** [comment]
@gdavison, I had a similar issue when I copied the user data from the acceptance tests.

I compared it to the user data in the EKS docs and adding a newline somewhere between the MIME tag solved this for me. Something is weird with parsing on the serverside IMO. I'll try to find the config I used.

**Comment by cohandv** [comment]
Hi @gdavison sorry I missed it. This is the user data: https://gist.github.com/cohandv/cf446ad946e54a3c35207efac8cde01c

Just for clarification this scenario occurs when you have created a nodegroup with a launch template using that user data. Then I changed the list of security groups (same user data) which required to create a new version of LT and update the existing nodegroup. Just then I got the error about user data mime which was not changed

**Comment by gdavison** [comment]
Interesting. Based on the error message and @DrFaust92's comment about server-side parsing above, I wonder if the AWS API is accepting a "bare" bash script on creation, but requires a MIME-multipart block on update. I'll do some testing on this.

**Comment by gdavison** [comment]
I've done some digging, and it looks like we've hit this error as well: https://github.com/terraform-providers/terraform-provider-aws/blob/374a49a5878e95a70e7c2a4f3cb88a6e68da8df0/aws/resource_aws_eks_node_group_test.go#L415

However, the [AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html#launch-template-user-data) for Launch Templates with EKS says:
> Amazon EC2 user data in launch templates that are used with managed node groups must be in the MIME multi-part archive format. This is because your user data is merged with Amazon EKS user data required for nodes to join the cluster.

It looks like the error here is that the AWS API accepts user data not in MIME multi-part format when creating an EKS node group, but checks on update.

**Comment by cohandv** [comment]
@gdavison yes this was it! After putting the user data in that format, it worked.

is there a way terraform validates this format? or is by desing that we rely on the provider API? 

Should we open a case to aws for this?

Thanks for your time helping me with this

**Comment by gdavison** [comment]
This would be a tricky one for Terraform to validate, since it's across resources. For most Launch Templates, the user data does not need to be in MIME multi-part. It's when the Launch Template is used to configure an EKS cluster that it adds that extra requirement. We don't currently support cross-resource validation, since it could lead to some challenging coupling.

Ideally, the AWS API would either return the error on Create or allow non-multi-part on Update. It could be helpful to open a support case with AWS.

**Comment by Skaronator** [comment]
I got the same issue. After a bit of try and error it seems like that gzip'ing the user_data cause the issue for me.

Setting gzip to `false` and everything is fine. Setting it to `true` I'll get `Error 1: Code: Unknown / Message: User data was not in the MIME multipart format.`. I guess the AWS server side validation doesn't work correctly.

```
data "cloudinit_config" "new" {
  gzip          = false
  base64_encode = true
  # https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html
  boundary = "==BOUNDARY=="

  part {
    content_type = "text/cloud-config"
    content      = file("${path.module}/worker/cloud-init.yaml")
  }
}
```
